waspc/src/Wasp/Generator/Valid/PackageJson.hs

```diff
@@ -12,163 +18,180 @@ import Wasp.Generator.DepVersions (prismaVersion, typescriptVersion)
+                ("@types/express", show expressTypesVersion)
+              ]
+        ]
+
+-- | Validates that an optional dependency is either not present, or present
+-- with the correct version.
+makeOptionalDepValidator :: DependencyType -> DependencySpecification -> V.Validator' P.PackageJson
+makeOptionalDepValidator depType dep@(pkgName, expectedPkgVersion) =
+  inDependency depType dep optionalVersionValidator
   where
-    validateRequiredWorkspaces Nothing = [missingWorkspacesError]
-    validateRequiredWorkspaces (Just definedWorkspacesList)
-      | NW.workspaceGlobs `S.isSubsetOf` definedWorkspaces = []
-      | otherwise = [makeWrongWorkspacesError definedWorkspaces]
-      where
-        definedWorkspaces = S.fromList definedWorkspacesList
-
-    makeWrongWorkspacesError definedWorkspaces =
-      GenericGeneratorError $
-        unwords
-          [ "Wasp requires package.json \"workspaces\" to include:",
-            showSet NW.workspaceGlobs ++ ".",
-            "You are missing:",
-            showSet (NW.workspaceGlobs `S.difference` definedWorkspaces) ++ "."
-          ]
-
-    missingWorkspacesError =
-      GenericGeneratorError $
-        unwords
-          [ "Wasp requires package.json \"workspaces\" to be present with the value and include values:",
-            showSet NW.workspaceGlobs ++ "."
-          ]
-
-    showSet = intercalate ", " . fmap show . S.toList
-
-validatePackageJsonDependency :: P.PackageJson -> PackageSpecification -> PackageRequirement -> [GeneratorError]
-validatePackageJsonDependency packageJson (packageName, expectedPackageVersion) requirement =
-  case maybePackageJsonDepedency of
-    Just actualPackageVersion ->
-      if actualPackageVersion == expectedPackageVersion
-        then []
-        else [incorrectPackageVersionErrorMessage]
-    Nothing ->
-      if isInWrongLocation requirement
-        then [wrongDependencyTypeErrorMessage]
-        else getMissingPackageError requirement
+    optionalVersionValidator :: V.Validator' (Maybe P.PackageVersion)
+    optionalVersionValidator actualVersion =
+      case (expectedPkgVersion ==) <$> actualVersion of
+        Just True -> V.success
+        Just False -> incorrectVersionError
+        Nothing -> V.success
+
+    incorrectVersionError =
+      V.failure $
+        "Wasp requires package "
+          ++ show pkgName
+          ++ " to be version "
+          ++ show expectedPkgVersion
+          ++ " if present."
+
+-- | Validates that a required dependency is present in the correct dependency
+-- list with the correct version. It shows an appropriate error message
+-- otherwise (with an explicit check for the case when the dependency is present
+-- in the opposite list -- runtime deps vs. devDeps).
+makeRequiredDepValidator :: DependencyType -> DependencySpecification -> V.Validator' P.PackageJson
+makeRequiredDepValidator depType dep@(pkgName, expectedPkgVersion) pkgJson =
+  whenS (inOppositeDepList notPresentValidator pkgJson) $
```

Comment:
Ok so I do have to admit at the end, I don't like this whenS too much, actual `and` sounds more reasonable in the way it communicates shortcircuiting, because we are all used to second thing not happening in first thing in `and` fails. Because with `when` it sounds like first thing is condiiton and second thing is THE thing, makes them non-equal, which is not the case here. So yeah I would probably vote for going back to `and`, but I know this is not a big thing + I am going against my previous self hah so I am ok with leaving it like this also, maybe not worth bothering, RAW.

Reply:
I created a nice `and` function, and this allowed me to remove the `Validator` and `Validator'` split, as now all the Validation contents are always `()`